include/rmm/mr/device/arena_memory_resource.hpp

```diff
@@ -0,0 +1,293 @@
+
+ private:
+  using global_arena = detail::arena::global_arena<Upstream>;
+  using arena        = detail::arena::arena<Upstream>;
+  using read_lock    = std::shared_lock<std::shared_timed_mutex>;
+  using write_lock   = std::lock_guard<std::shared_timed_mutex>;
+
+  /**
+   * @brief Allocates memory of size at least `bytes`.
+   *
+   * The returned pointer has at least 256-byte alignment.
+   *
+   * @throws `std::bad_alloc` if the requested allocation could not be fulfilled.
+   *
+   * @param bytes The size in bytes of the allocation.
+   * @param stream The stream to associate this allocation with.
+   * @return void* Pointer to the newly allocated memory.
+   */
+  void* do_allocate(std::size_t bytes, cudaStream_t stream) override
+  {
+    if (bytes <= 0) return nullptr;
+
+    bytes = detail::arena::align_up(bytes);
+
+    if (arena::handles_size(bytes)) {
+      return get_arena(stream).allocate(bytes);
+    } else {
+      return global_arena_.allocate(bytes);
+    }
+  }
+
+  /**
+   * @brief Deallocate memory pointed to by `p`.
+   *
+   * @param p Pointer to be deallocated.
+   * @param bytes The size in bytes of the allocation. This must be equal to the
+   * value of `bytes` that was passed to the `allocate` call that returned `p`.
+   * @param stream Stream on which to perform deallocation.
+   */
+  void do_deallocate(void* p, std::size_t bytes, cudaStream_t stream) override
+  {
+    if (p == nullptr || bytes <= 0) return;
+
+    bytes = detail::arena::align_up(bytes);
+    if (arena::handles_size(bytes)) {
+      if (!get_arena(stream).deallocate(p, bytes, stream)) {
+        deallocate_from_other_arena(p, bytes, stream);
+      }
+    } else {
+      RMM_ASSERT_CUDA_SUCCESS(cudaStreamSynchronize(stream));
+      global_arena_.deallocate(p, bytes);
+    }
+  }
+
+  /**
+   * @brief Deallocate memory pointed to by `p` that was allocated in a different arena.
+   *
+   * @param p Pointer to be deallocated.
+   * @param bytes The size in bytes of the allocation. This must be equal to the
+   * value of `bytes` that was passed to the `allocate` call that returned `p`.
+   * @param stream Stream on which to perform deallocation.
+   */
+  void deallocate_from_other_arena(void* p, std::size_t bytes, cudaStream_t stream)
+  {
+    RMM_ASSERT_CUDA_SUCCESS(cudaStreamSynchronize(stream));
+
+    read_lock lock(mtx_);
+
+    if (use_per_thread_arena(stream)) {
+      auto const id = std::this_thread::get_id();
+      for (auto& kv : thread_arenas_) {
+        // Check the per-thread arena if it does not belong to the current thread, and return if the
+        // pointer is found.
+        if (kv.first != id && kv.second.deallocate(p, bytes)) return;
+      }
+    } else {
+      for (auto& kv : stream_arenas_) {
+        // Check the per-stream arena if it does not belong to the current stream, and return if the
+        // pointer is found.
+        if (kv.first != stream && kv.second.deallocate(p, bytes)) return;
+      }
```

Comment:
Same here. Iterating a `unordered_map` when a direct lookup could be used.

Reply:
See above.